<?xml version="1.0"?>
<launch>
       
  <arg name="map_file" default="$(find maker_robot)/maps/layout.yaml"/>
  <arg name="model" default="$(find maker_robot)/urdf/my_robot_car_4.xacro"/>
  
  <rosparam file="$(find maker_robot)/config/controllers.yaml" command="load"/>
  <rosparam file="$(find maker_robot)/config/joint_limits.yaml" command="load"/>

   
  <param name="robot_description" command="$(find xacro)/xacro $(arg model)" />
  
  <node name="robot_hardware_interface" pkg="maker_robot" type="mobile_robot_hardware_interface" output="screen">
    <remap from="/mobile_robot/mobile_base_controller/cmd_vel" to="/cmd_vel"/>
      
     <remap from="/mobile_robot/mobile_base_controller/odom"   to="/odom"/>
        
   </node>

  <node name="subscriber_py" pkg="maker_robot" type="joint_receive_from_arduino.py" output="screen"/> 
  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" >
  </node>

 
 
  
  
  <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen"
        args="
            /mobile_robot/joints_update
            /mobile_robot/mobile_base_controller" >
  </node>
 <!--
  <node name="telop" pkg="teleop_twist_keyboard" type="teleop_twist_keyboard.py"> -->
		<!--<remap from="/cmd_vel" to="/mobile_robot/mobile_base_controller/cmd_vel"/> -->
 <!--  </node> -->
  
   <node pkg="rosserial_python" type="serial_node.py" name="serial_node">
    <param name="port" value="/dev/ttyACM0"/>
    <param name="baud" value="115200"/>
  </node>

<include file="$(find rplidar_ros)/launch/rplidar.launch" />

  <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_file)" >
   </node> 

  <!-- Place map frame at odometry frame -->
  

  <!-- AMCL -->
  <node pkg="amcl" type="amcl" name="amcl" output="screen">
    
    <param name="odom_frame_id" value="odom"/>
    <param name="odom_model_type" value="diff"/>
    <param name="base_frame_id" value="base_link"/>
    <param name="update_min_d" value="0.5"/>
    <param name="update_min_a" value="1.0"/>

    <param name="min_particles"             value="500"/>
    <param name="max_particles"             value="3000"/>
    <param name="kld_err"                   value="0.02"/>
    <param name="resample_interval"         value="1"/>
    <param name="transform_tolerance"       value="0.5"/>
    <param name="recovery_alpha_slow"       value="0.00"/>
    <param name="recovery_alpha_fast"       value="0.00"/>
    <param name="gui_publish_rate"          value="50.0"/>

    <param name="laser_max_range"           value="3.5"/>
    <param name="laser_max_beams"           value="180"/>
    <param name="laser_z_hit"               value="0.5"/>
    <param name="laser_z_short"             value="0.05"/>
    <param name="laser_z_max"               value="0.05"/>
    <param name="laser_z_rand"              value="0.5"/>
    <param name="laser_sigma_hit"           value="0.2"/>
    <param name="laser_lambda_short"        value="0.1"/>
    <param name="laser_likelihood_max_dist" value="2.0"/>
    <param name="laser_model_type"          value="likelihood_field"/>

    <param name="odom_alpha1"               value="0.1"/>
    <param name="odom_alpha2"               value="0.1"/>
    <param name="odom_alpha3"               value="0.1"/>
    <param name="odom_alpha4"               value="0.1"/>

  </node>
  
  <!-- Arguments -->
  <arg name="cmd_vel_topic" default="/cmd_vel" />
  <arg name="odom_topic" default="odom" />
  <arg name="move_forward_only" default="false"/>

  <!-- move_base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
    <rosparam file="$(find maker_robot)/param/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find maker_robot)/param/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find maker_robot)/param/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find maker_robot)/param/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find maker_robot)/param/move_base_params.yaml" command="load" />
    <rosparam file="$(find maker_robot)/param/dwa_local_planner_params.yaml" command="load" />
    <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
    <remap from="odom" to="$(arg odom_topic)"/>
    <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)" />
  </node>
<!--
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find maker_robot)/config/rviz/nav_config2.rviz"/>
 -->
</launch>
